src/tree_walker.cpp

```diff
@@ -0,0 +1,179 @@
+#include "tree_walker.h"
+#include <iostream>
+
+#include <string> //feel free to delete later
+
+using namespace smt;
+using namespace std;
+
+namespace smt
+{
+
+//DELETE
+std::string vectorToString(vector<int> v){
+  std::string s;
+  for (int i:v){
+    s += std::to_string(i);
+    s += ", ";
+  }
+  return s;
+}
+
+pair<Term, vector<int>> TreeWalker::visit(Term & node)
+{
+  //iterates over children, for tracking which child on & path
+  int child_no;
+  //path for parts of the formula
+  vector<int> tree_path;
+
+  if (clear_cache_)
+  {
+    cache_.clear();
+
+    if (ext_cache_)
+    {
+      ext_cache_->clear();
+    }
+  }
+
+  pair <Term, vector<int>> out;
+  out.first = node;
+  out.second = tree_path;
+  if (query_cache(node, out))
+  {
+    // cache hit
+    return out;
+  }
+
+  //tree_path currently empty, process top node
+  visit_term(node, node, tree_path);
+  TermPairVec to_visit;
+  pair<Term, int> p1 (node, -1);
+  to_visit.push_back(p1);
+  child_no = 0;
+  for (auto ttt : node){
```

Comment:
We have a very similar loop inside the main loop. So please explain (in detail) why do we do this twice, why can't we just add only in the loop.